### PR TITLE
HPCC-23215 Start experimenting with code to read s3 buckets

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
@jakesmith @richardkchapman @jpmcmu 
FYI this is my initial PR that supports reading from S3.  I need to move the AWS libraries and do more work before anything like ready for review, but here so you can have a look/comment.

File: common/remote/hooks/s3/s3file.cpp [hunk 1]
```diff
@@ -0,0 +1,744 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2019 HPCC Systems®.
```
Comment (richardkchapman) [inline common/remote/hooks/s3/s3file.cpp]: 2020

File: common/remote/hooks/s3/s3file.cpp [hunk 2]
```diff
@@ -0,0 +1,744 @@
...
+        modifiedTime->set((time_t)(modifiedMsTime / 1000));
+    }
+    if (accessedTime)
+        accessedTime->clear();
+    return false;
+}
+
+void S3File::readBlob(Aws::S3::Model::GetObjectOutcome & readResult, FileIOStats & stats, offset_t from, offset_t length)
+{
+    Aws::S3::S3Client s3_client = getAwsClient();
```
Comment (richardkchapman) [inline common/remote/hooks/s3/s3file.cpp]: Do at open time?

File: common/remote/hooks/s3/s3file.cpp [hunk 3]
```diff
@@ -0,0 +1,744 @@
...
+    {
+        StringBuffer range;
+        range.append("bytes=").append(from).append("-");
+        if (length != unknownFileSize)
+            range.append(from + length - 1);
+        object_request.SetRange(Aws::String(range));
+    }
+
+    // Get the object
+    CCycleTimer timer;
```
Comment (richardkchapman) [inline common/remote/hooks/s3/s3file.cpp]: May need a thread reading ahead?

File: common/remote/hooks/s3/s3file.cpp [hunk 4]
```diff
@@ -0,0 +1,744 @@
...
+    bool haveMeta = false;
+    bool isDir = false;
+    bool fileExists = false;
+    int64_t modifiedMsTime = 0;
+    SpinLock lock;
+};
+
+//---------------------------------------------------------------------------------------------------------------------
+
+unsigned __int64 FileIOStats::getStatistic(StatisticKind kind)
```
Comment (richardkchapman) [inline common/remote/hooks/s3/s3file.cpp]: Move to jlib

File: common/remote/hooks/s3/s3file.cpp [hunk 5]
```diff
@@ -0,0 +1,744 @@
...
+    return true;
+}
+MODULE_EXIT()
+{
+    Aws::ShutdownAPI(options);
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+
+class FileIOStats
```
Comment (richardkchapman) [inline common/remote/hooks/s3/s3file.cpp]: Move to jlib

File: common/remote/hooks/s3/s3file.cpp [hunk 6]
```diff
@@ -0,0 +1,744 @@
...
+/*
+ * S3 questions:
+ *
+ * Where would we store access id/secrets?
+ * What is the latency on file access?  ~200ms.
+ * What is the cost of accessing the data?  $0.4/million GET requests (including requesting meta)
+ * How to you efficiently page results from a S3 bucket?  You can perform a range get, but you'll need to pay for each call.
+ * You can get the length of an object using HeadObject..getContentLength, but you will get charged - so for small files it is better to just get it.
+ *     Probably best to request the first 10Mb, and if that returns 10Mb then request the size information.
+ * S3 does not support partial writes, so we would need to create a file locally and then write it in one go.  Not sure how you do that if it doesn't fit into memory...
```
Comment (richardkchapman) [inline common/remote/hooks/s3/s3file.cpp]: Oh yes it does!

**Comment by ghalliday** [comment]
@richardkchapman @jakesmith @jpmcmu 
For initial review.  It is a very provisional implementation, but some of the changes are worth getting in (e.g., the changes to isAbsoluteFilename())

My sample command line:
`eclcc s3://gchtestingbucket/toplevel/counting.ecl -E -o s3://gchtestingbucket/myarchive.txt
`
with environment variables:
```
AWS_ACCESS_KEY_ID=....
AWS_SECRET_ACCESS_KEY=....
HPCC_FILEHOOKS_PATH=/home/gavin/build/Debug/libs/libs3file.so
```

File: common/remote/hooks/s3/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,50 @@
+################################################################################
+#    HPCC SYSTEMS software Copyright (C) 2019 HPCC Systems®.
```
Comment (richardkchapman) [inline common/remote/hooks/s3/CMakeLists.txt]: 2020 ?

**Review by richardkchapman — APPROVED** [review]
Minor comments

**Comment by richardkchapman** [comment]
@jakesmith

File: common/remote/hooks/s3/s3file.cpp [hunk 8]
```diff
@@ -0,0 +1,721 @@
...
+    ensureMetaData();
+    if (createTime)
+        createTime->clear();
+    if (modifiedTime)
+    {
+        modifiedTime->clear();
+        modifiedTime->set((time_t)(modifiedMsTime / 1000));
+    }
+    if (accessedTime)
+        accessedTime->clear();
```
Comment (jakesmith) [inline common/remote/hooks/s3/s3file.cpp]: I assume create/accessed time isn't available.
If unknown, would it be better to set create time to modified time

**Comment by ghalliday** [comment]
@jakesmith see changes in new commit.

**Comment by jakesmith** [comment]
@ghalliday - changes look good.

**Pull request merged**